File: Java/프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.HashMap;
import java.util.ArrayList;

class Solution {
    private static HashMap<String, Integer> termsLimit = new HashMap<>();
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        splitTermsLimit(terms);
        
        int todayTotalDays = dateToDays(today);
        
        ArrayList<Integer> expiredList = new ArrayList<>();
        
        for (int i = 0; i < privacies.length; i++) {
            String[] splitPrivacy = privacies[i].split(" ");
            String dateStr = splitPrivacy[0];
            String termType = splitPrivacy[1];
            
            int collectedDays = dateToDays(dateStr);
            int limitMonth = termsLimit.get(termType);
            
            if (collectedDays + (limitMonth * 28) <= todayTotalDays) {
                expiredList.add(i + 1);
            }
        }
        
        return expiredList.stream().mapToInt(Integer::intValue).toArray();
    }
    
    private void splitTermsLimit(String[] terms) {
        for (String term: terms) {
            String[] termSplited = term.split(" ");
            termsLimit.put(termSplited[0], Integer.parseInt(termSplited[1]));
        }
    }
    
    private int dateToDays(String date) {
        String[] dateSplited = date.split("\\.");
        int year = Integer.parseInt(dateSplited[0]);
        int month = Integer.parseInt(dateSplited[1]);
        int day = Integer.parseInt(dateSplited[2]);
        return (year * 12 * 28) + (month *28) + day;
    }
}

File: Swift/프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.swift
import Foundation

var termsLimit: [String : Int] = [:]

func solution(_ today:String, _ terms:[String], _ privacies:[String]) -> [Int] {
    splitTermsLimit(terms)
    let todayTotalDays = dateToDays(today)
    var expiredList: [Int] = []
    
    for i in 0..<privacies.count {
        let splitPrivacy = privacies[i].split(separator: " ").map({ String($0) })
        let dateStr = String(splitPrivacy[0])
        let termType = String(splitPrivacy[1])
        let collectedDays = dateToDays(dateStr)
        let limitMonth = termsLimit[termType] ?? 0
        
        if (collectedDays + (limitMonth * 28)) <= todayTotalDays {
            expiredList.append(i+1)
        }
    }
    return expiredList
}

func splitTermsLimit(_ terms: [String]) {
    for term in terms {
        let termSplited = term.split(separator: " ").map({ String($0) })
        termsLimit[termSplited[0]] = Int(termSplited[1])!
    }
}

func dateToDays(_ date: String) -> Int {
    let dateSplited = date.split(separator: ".")
    let year = Int(dateSplited[0])!
    let month = Int(dateSplited[1])!
    let day = Int(dateSplited[2])!
    
    return (year * 12 * 28) + (month * 28) + day
}